File: app/src/main/java/com/zattoo/movies/ui/home/HomeFragment.java
package com.zattoo.movies.ui.home;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import com.zattoo.movies.MoviesApp;
import com.zattoo.movies.R;
import com.zattoo.movies.utils.DataUtilsKt;
import com.zattoo.movies.utils.NetworkUtils;
import com.zattoo.movies.data.home.Movie;
import com.zattoo.movies.data.MovieListEntity;
import com.zattoo.movies.data.MovieListOffers;
import com.zattoo.movies.data.MovieService;
import com.zattoo.movies.databinding.FragmentHomeBinding;

import java.util.List;

import javax.inject.Inject;

import retrofit2.Response;

public class HomeFragment extends Fragment {

    final static int ANIMATION_DURATION = 1000;

    @Inject
    HomeAdapter adapter;
    @Inject
    NetworkUtils networkUtils;
    @Inject
    MovieService movieService;

    private FragmentHomeBinding binding;
    private Result.Success success;

    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            ViewGroup container,
            Bundle savedInstanceState
    ) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_home, container, false);
        binding.setLifecycleOwner(this);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ((MoviesApp) getContext().getApplicationContext()).getApplicationComponent()
                .inject(this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initUiElements();
    }

    @Override
    public void onStart() {
        super.onStart();
        fetchData();
        handleNetwork();
    }

    private void initUiElements() {
        adapter = new HomeAdapter();
        binding.recyclerView.setAdapter(adapter);
        binding.swipeRefreshLayout.setOnRefreshListener(() -> {
            fetchData();
        });
    }

    private void handleLoading() {
        showLoading(true);
    }

    private void handleResults(Result.Success result) {
        showLoading(false);
        if (result.movies.isEmpty()) {
            handleError();
        } else {
            this.success = result;
            adapter.setList(this);
        }
    }

    private void handleError(final String error) {
        handleError(false, error);
    }

    private void handleError() {
        handleError(true, "");
    }

    private void handleError(final Boolean isEmptyList, final String error) {
        showLoading(false);
        if (isEmptyList) {
            final String errorMessage = getString(R.string.empty_list);
            showEmptyList(errorMessage);
        } else {
            showError(error);
        }
    }

    private void showEmptyList(String message) {
        binding.recyclerView.setVisibility(View.GONE);
        binding.emptyView.setVisibility(View.VISIBLE);
        binding.emptyView.setText(message);
    }

    private void showError(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
    }

    private void showLoading(boolean isLoading) {
        binding.swipeRefreshLayout.setRefreshing(isLoading);
    }

    private void handleNetwork() {
        networkUtils.getNetworkLiveData().observe(this, isConnected -> {
            if (!isConnected) {
                binding.textViewNetworkStatus.setText(getString(R.string.text_no_connectivity));
                binding.networkStatusLayout.setVisibility(View.VISIBLE);
                binding.networkStatusLayout.setBackgroundColor(ResourcesCompat.getColor(
                        getResources(),
                        R.color.colorStatusNotConnected,
                        null
                ));
                binding.swipeRefreshLayout.setRefreshing(false);
            } else {
                fetchData();
                binding.textViewNetworkStatus.setText(getString(R.string.text_connectivity));
                binding.networkStatusLayout.setBackgroundColor(ResourcesCompat.getColor(
                        getResources(), R.color.colorStatusConnected, null
                ));

                binding.networkStatusLayout.animate().alpha(1f)
                        .setStartDelay(ANIMATION_DURATION)
                        .setDuration(ANIMATION_DURATION)
                        .setListener(new AnimatorListenerAdapter() {
                            @Override
                            public void onAnimationEnd(Animator animation) {
                                super.onAnimationEnd(animation);
                                binding.networkStatusLayout.setVisibility(View.GONE);
                            }
                        });
            }
        });
    }

    private void fetchData() {
        new FetchDataAsyncTask().execute();
    }

    List<Movie> getMovies() {
        return success.movies;
    }

    class FetchDataAsyncTask extends AsyncTask<Void, Void, Result> {
        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            handleLoading();
        }

        @Override
        protected Result doInBackground(Void... voids) {
            try {
                Response<MovieListEntity> responseMovieDataList = DataUtilsKt.fetchMovieList(movieService);
                Response<MovieListOffers> responseMovieListOffers = DataUtilsKt.fetchMovieListOffers(movieService);
                if (responseMovieDataList.isSuccessful() && responseMovieListOffers.isSuccessful()) {
                    return new Result.Success(DataUtilsKt.createMovies(
                            responseMovieDataList.body().getMovie_data(),
                            responseMovieListOffers.body()));
                } else {
                    return new Result.Failure("fetchMovieList or fetchMovieListOffers failed");
                }

            } catch (Exception exception) {
                return new Result.Failure(exception.getMessage());
            }
        }

        @Override
        protected void onPostExecute(Result result) {
            super.onPostExecute(result);
            if (result instanceof Result.Success) {
                final Result.Success success = (Result.Success) result;
                handleResults(success);
            } else if (result instanceof Result.Failure) {
                final Result.Failure failure = (Result.Failure) result;
                handleError(failure.message);
            }
        }
    }
}

File: app/src/main/java/com/zattoo/movies/ui/home/HomeFragment.kt
package com.zattoo.movies.ui.home

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import com.zattoo.movies.R
import com.zattoo.movies.data.home.Movie
import com.zattoo.movies.databinding.FragmentHomeBinding
import com.zattoo.movies.utils.NetworkUtils
import com.zattoo.movies.utils.Resource
import com.zattoo.movies.viewmodel.MovieViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class HomeFragment: Fragment() {

    @Inject
    lateinit var adapter: HomeAdapter

    @Inject
    lateinit var networkUtils: NetworkUtils

    private var _binding: FragmentHomeBinding? = null

    private val binding get() = _binding!!

    private val viewModel: MovieViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false).apply {
            lifecycleOwner = viewLifecycleOwner
        }
        handleResults()
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initUiElements()
    }

    override fun onStart() {
        super.onStart()
        handleNetwork()
    }

    private fun initUiElements() {
        adapter = HomeAdapter()
        binding.recyclerView.adapter = adapter
        binding.swipeRefreshLayout.setOnRefreshListener { viewModel.refresh() }
    }

    private fun handleResults(movies: List<Movie>) {
        showLoading(false)
        if (movies.isEmpty()) {
            handleError()
        } else {
            adapter.setList(movies)
        }
    }

    private fun handleError(error: String) {
        handleError(false, error)
    }

    private fun handleError() {
        handleError(true, "")
    }

    private fun handleError(isEmptyList: Boolean, error: String) {
        showLoading(false)
        if (isEmptyList) {
            val errorMessage = getString(R.string.empty_list)
            showEmptyList(errorMessage)
        } else {
            showError(error)
        }
    }

    private fun showEmptyList(message: String) {
        binding.recyclerView.visibility = View.GONE
        binding.emptyView.visibility = View.VISIBLE
        binding.emptyView.text = message
    }

    private fun showError(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }

    private fun showLoading(isLoading: Boolean) {
        binding.swipeRefreshLayout.isRefreshing = isLoading
    }

    private fun handleNetwork() {
        networkUtils.getNetworkLiveData().observe(this) { isConnected: Boolean ->
            if (!isConnected) {
                binding.textViewNetworkStatus.text = getString(R.string.text_no_connectivity)
                binding.networkStatusLayout.visibility = View.VISIBLE
                binding.networkStatusLayout.setBackgroundColor(
                    ResourcesCompat.getColor(
                        resources,
                        R.color.colorStatusNotConnected,
                        null
                    )
                )
                binding.swipeRefreshLayout.isRefreshing = false
            } else {
                viewModel.refresh()
                binding.textViewNetworkStatus.text = getString(R.string.text_connectivity)
                binding.networkStatusLayout.setBackgroundColor(
                    ResourcesCompat.getColor(
                        resources, R.color.colorStatusConnected, null
                    )
                )
                binding.networkStatusLayout.animate().alpha(1f)
                    .setStartDelay(ANIMATION_DURATION.toLong())
                    .setDuration(ANIMATION_DURATION.toLong())
                    .setListener(object : AnimatorListenerAdapter() {
                        override fun onAnimationEnd(animation: Animator) {
                            super.onAnimationEnd(animation)
                            binding.networkStatusLayout.visibility = View.GONE
                        }
                    })
            }
        }
    }

    private fun handleResults() {
        lifecycleScope.launch {
            viewModel.stateFlow.collect {
                when (it) {
                    is Resource.Success -> handleResults(it.data)
                    is Resource.Error -> handleError(it.message)
                    is Resource.Loading -> showLoading(true)
                }
            }
        }
    }

}

private const val ANIMATION_DURATION = 1000